src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3731,119 +3730,65 @@ private BoundExpression BindConditionalOperator(ConditionalExpressionSyntax node
                 CheckFeatureAvailability(node, MessageID.IDS_FeatureRefConditional, diagnostics);
             }
 
-            BoundExpression condition = BindBooleanExpression(node.Condition, diagnostics);
+            return isRef ? BindRefConditionalOperator(node, whenTrue, whenFalse, diagnostics) : BindValueConditionalOperator(node, whenTrue, whenFalse, diagnostics);
+        }
+
+        private BoundExpression BindValueConditionalOperator(ConditionalExpressionSyntax node, ExpressionSyntax whenTrue, ExpressionSyntax whenFalse, DiagnosticBag diagnostics)
+        {
+            ErrorCode noCommonTypeError = 0;
 
-            var valKind = BindValueKind.RValue;
-            if (isRef)
+            BoundExpression condition = BindBooleanExpression(node.Condition, diagnostics);
+            BoundExpression trueExpr = BindValue(whenTrue, diagnostics, BindValueKind.RValue);
+            BoundExpression falseExpr = BindValue(whenFalse, diagnostics, BindValueKind.RValue);
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            TypeSymbol type = BestTypeInferrer.InferBestTypeForConditionalOperator(trueExpr, falseExpr, this.Conversions, out bool hadMultipleCandidates, ref useSiteDiagnostics);
+            diagnostics.Add(node, useSiteDiagnostics);
+            bool targetTyped;
+            if (type is null)
             {
-                valKind |= BindValueKind.RefersToLocation;
+                noCommonTypeError = hadMultipleCandidates ? ErrorCode.ERR_AmbigQM : ErrorCode.ERR_InvalidQM;
+                targetTyped = Compilation.LanguageVersion >= MessageID.IDS_FeatureTargetTypedConditional.RequiredVersion();
```

Comment:
This is somewhat unfortunate, as we're not going to provide a langversion diagnostic. We should consider always target-typing, and just issuing an error if the language version isn't correct.

Reply:
The diagnostics for the error cases do not work out nicely that way.

---
In reply to: [410417426](https://github.com/dotnet/roslyn/pull/40604#discussion_r410417426) [](ancestors = 410417426)